### DISCUSSION TITLE
브라우저 이벤트 루프

### DISCUSSION THREAD

**Discussion by greenblues1190** [woowacourse-study/2022-woowahan-bansanghwe#138] [JavaScript]
자바스크립트는 싱글 스레드 기반의 언어이다. 이러한 자바스크립트 코드를 실행하는 브라우저가 비동기 로직을 처리하고 동시성을 지원하기 위해 채택한 런타임 모델이 이벤트 루프이다.

## 콜 스택

> 싱글 스레드 == 하나의 콜 스택 == 한번에 하나의 일을 처리

자바스크립트 엔진은 하나의 콜 스택에서 모든 명령을 처리한다. 함수가 호출되면 콜 스택에 쌓인다. 함수 내에서 함수가 호출되는 경우도 마찬가지이다. 자바스크립트는 콜 스택에서 명령을 하나씩 꺼내 실행한다. 이는 모든 콜 스택이 빌 때까지 유효하다. 웹 개발을 하며 만나는 `RangeError: Maximum call stack size exceeded`는 콜 스택이 가득 차 더 이상 명령을 넣을 수 없을 때 발생하는 에러이다.

자바스크립트 엔진은 하나의 콜 스택에서 모든 명령을 처리한다고 했다. 그런데 엄청나게 느린 명령이 콜 스택에 존재한다면 어떻게 될까? 자바스크립트 엔진은 콜 스택이 빌 때까지 계속해서 명령을 처리하기 때문에 해당 명령이 처리될 때까지 기다리게 된다. 그 동안 자바스크립트 엔진은 어떤 사용자 인터렉션도 처리하지 않고 화면을 업데이트하지도 않는다. 브라우저의 메인 스레드가 Blocking된 것이다.

그런데 특정 명령어를 만나면 다른 동작을 한다. 에를 들어 setTimeout과 같은 것들이 그렇다. 

```javascript
console.log(1);
setTimeout(() => { console.log(2); }, 0);
console.log(3);
```

위의 코드를 실행하면 다음과 같은 결과가 출력된다.

```shell
1
3
2
```

## 큐

이와 같은 현상이 발생하는 이유는 브라우저가 단순히 자바스크립트 런타임이 아니기 때문이다. 브라우저는 Web API를 지원한다. 자바스크립트가 실행되는 런타임 환경에 존재하는 별도의 API다. 다시 위의 코드를 보자.

1. 스택에 각 명령이 쌓인다.
2. 1을 출력하고 `console.log(1);`를 지운다.
3. 브라우저가 Web API를 이용해 타이머를 동작시키고 `setTimeout(() => { console.log(2); }, 0);`를 지운다.
4. 3을 출력하고 `console.log(3);`를 지운다.

이제 모든 콜 스택이 비었다. Web API에서 실행되고 있는 타이머는 0ms, 즉 바로 종료된다. Web API는 동작을 완료하면 콜백을 **태스크 큐**에 집어넣는다. 태스크 큐는 실행되어야 할 태스크들이 저장되는 장소이다. 여기서 이벤트 루프가 등장한다.

## 이벤트 루프

이벤트 루프는 콜 스택과 태스크 큐의 작업을 관리하는 역할을 한다. 이벤트 루프는 콜 스택이 비어있으면 태스크 큐에서 작업을 꺼내 콜 스택에 넣어준다. 자바스크립트 엔진은 콜 스택에 작업이 있으므로 `console.log(2);`를 실행한다. 따라서 1 -> 2 -> 3의 순서대로 콘솔 로그가 출력되는 것이다.

## 참고

- https://tecoble.techcourse.co.kr/post/2021-08-28-event-loop/
- https://medium.com/@francesco_rizzi/javascript-main-thread-dissected-43c85fce7e23
- https://meetup.toast.com/posts/89
- https://www.youtube.com/watch?v=YpQTeIqjC4o
- https://www.youtube.com/watch?v=8aGhZQkoFbQ

**Comment by al-bur** [comment]
자기소개

**Comment by al-bur** [comment]
협업을 지향하는 이 제일 본인은 잘 나타내는 키워드?

**Reply by al-bur** [reply]
어떤 노력을 하고있죠?

**Reply by al-bur** [reply]
본인은 팀에서 어떤 역할

**Reply by al-bur** [reply]
팀에서 가장 어려웠던 경험

**Comment by al-bur** [comment]
이벤트 루프를 간단하게 설명해주세요

**Comment by al-bur** [comment]
동시성이 뭔가요?

**Comment by al-bur** [comment]
http1 vs http2

**Comment by al-bur** [comment]
https는 어떤 부분을 보완?

**Comment by al-bur** [comment]
TLS handshaking

**Comment by al-bur** [comment]
콜과 스택의 차이

**Reply by al-bur** [reply]
개발할 때 언제 해당 자료구조를 사용하나요?

**Comment by JUDONGHYEOK** [comment]
리액트에서 버추얼돔을 채택한 이유?

**Comment by JUDONGHYEOK** [comment]
최근에 나온 라이브러리들은 왜 버추얼돔을 안쓰는지?

**Comment by JUDONGHYEOK** [comment]
https://www.google.com/을 쳤을 때 일어나는 일
